File: apps/web/static/js/modules/utils/validators.js
export function isRequired(value) {
  return typeof value === "string" && value.trim().length > 0;
}

export function isEmail(value) {
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(String(value).trim());
}

export function isUrl(value) {
  try {
    new URL(value);
    return true;
  } catch {
    return false;
  }
}

export function isMinLength(value, min) {
  return String(value).length >= min;
}

export function isMaxLength(value, max) {
  return String(value).length <= max;
}

// Strong password: 12+ chars, upper, lower, digit, symbol.
export function isStrongPassword(value) {
  const v = String(value);
  return (
    v.length >= 12 && /[A-Z]/.test(v) && /[a-z]/.test(v) && /\d/.test(v) && /[^A-Za-z0-9]/.test(v)
  );
}

// App name HTML: only <b> <strong> <i> <em> <u> <sup> <sub> are allowed, all tags must be balanced.
const _ALLOWED_APP_NAME_TAGS = ["b", "strong", "i", "em", "u", "sup", "sub"];

export function isValidAppNameHtml(value) {
  const str = String(value ?? "");
  if (!str.includes("<")) return true;

  const tagRe = /<\/?([a-zA-Z][a-zA-Z0-9]*)(?:\s[^>]*)?\s*\/?>/g;
  const stack = [];
  let match;

  while ((match = tagRe.exec(str)) !== null) {
    const full = match[0];
    const name = match[1].toLowerCase();
    if (!_ALLOWED_APP_NAME_TAGS.includes(name)) return false;
    if (full.startsWith("</")) {
      if (stack.length === 0 || stack[stack.length - 1] !== name) return false;
      stack.pop();
    } else if (!full.endsWith("/>")) {
      stack.push(name);
    }
  }

  return stack.length === 0;
}

// AWS Access Key ID: exactly 20 uppercase alphanumeric characters.
export function isAwsAccessKeyId(value) {
  return /^[A-Z0-9]{20}$/.test(String(value).trim());
}

// AWS Secret Access Key: exactly 40 base64 characters.
export function isAwsSecretAccessKey(value) {
  return /^[A-Za-z0-9/+=]{40}$/.test(String(value).trim());
}

// AWS region: e.g. eu-west-1, us-east-1, ap-southeast-2.
export function isAwsRegion(value) {
  return /^[a-z]{2}-[a-z]+-\d+$/.test(String(value).trim());
}

// Fernet key: 44-char URL-safe base64 (43 data chars + 1 trailing '=').
export function isFernetKey(value) {
  return /^[A-Za-z0-9_-]{43}=$/.test(String(value).trim());
}

// X.509 certificate body (no -----BEGIN/END CERTIFICATE----- header or footer).
export function isX509Cert(value) {
  const stripped = String(value).replace(/[\s\r\n]/g, "");
  if (stripped.includes("-----")) return false;
  return stripped.length >= 50 && /^[A-Za-z0-9+/]+=*$/.test(stripped);
}

// S3 bucket ARN: arn:aws:s3:::<bucket-name> (bucket 3–63 chars, lowercase alphanumeric, hyphens, dots).
export function isS3Arn(value) {
  return /^arn:aws:s3:::[a-z0-9][a-z0-9.-]{1,61}[a-z0-9]$/.test(String(value).trim());
}

File: apps/web/apps/setup/validators.py
import re

from django.core.exceptions import ValidationError
from django.core.validators import RegexValidator

AWS_ACCESS_KEY_ID_VALIDATOR = RegexValidator(
    regex=r"^[A-Z0-9]{20}$",
    message="Enter a valid AWS Access Key ID (20 uppercase alphanumeric characters).",
    code="invalid_aws_access_key_id",
)

AWS_SECRET_ACCESS_KEY_VALIDATOR = RegexValidator(
    regex=r"^[A-Za-z0-9/+=]{40}$",
    message="Enter a valid AWS Secret Access Key (40 characters).",
    code="invalid_aws_secret_access_key",
)

AWS_REGION_VALIDATOR = RegexValidator(
    regex=r"^[a-z]{2}-[a-z]+-\d+$",
    message="Enter a valid AWS region (e.g. eu-west-1, us-east-1, ap-southeast-2).",
    code="invalid_aws_region",
)

S3_ARN_VALIDATOR = RegexValidator(
    regex=r"^arn:aws:s3:::[a-z0-9][a-z0-9.\-]{1,61}[a-z0-9]$",
    message="Enter a valid S3 bucket ARN (e.g. arn:aws:s3:::my-bucket).",
    code="invalid_s3_arn",
)


def validate_x509_cert(value: str) -> None:
    """
    Validates a bare X.509 certificate body
    (no -----BEGIN/END CERTIFICATE----- markers).
    """
    if not value:
        return
    stripped = re.sub(r"[\s\r\n]", "", value)
    if "-----" in stripped:
        raise ValidationError(
            (
                "Do not include the -----BEGIN CERTIFICATE----- header or footer — "
                "paste only the base64 body."
            ),
            code="invalid_x509_cert",
        )
    if len(stripped) < 50 or not re.fullmatch(r"[A-Za-z0-9+/]+=*", stripped):
        raise ValidationError(
            "Enter a valid base64-encoded X.509 certificate body.",
            code="invalid_x509_cert",
        )
